### ISSUE TITLE
Running many irm in parallel when using Oracle DB causes some to fail (does not exist)

### ISSUE THREAD

**Issue by keithj** [irods/irods#3565]
iRODS 4.1.10, icommands 4.1.10

Starting with a collection (`trash/orphan`) containing 30k data objects I created a file of commands like this:
````
irm /seq/trash/orphan/file1
irm /seq/trash/orphan/file2

etc
````
I passed this file to GNU parallel to run 20 clients concurrently. When all the commands had been executed, many files were left undeleted and I observed many errors of like:

````
ERROR: rmUtil: srcPath /seq/trash/orphan/file1 does not exist
````

I tested a few of these files to confirm that they did exist (with `ils`) and could be removed with `irm`. I then created a new command list as above, containing only the remaining files to be deleted. Running these in parallel gave the same behaviour, again leaving a subset of the files undeleted. I had to repeat the process several times to get all the files to delete.

(Reason for doing this in parallel - `irm -r` which I tried first was too slow (possibly deleting 30k files sequentially?)

**Comment by tempoz** [comment]
Is it correct to assume you have trash off (`msiNoTrashCan`) for this, as otherwise irm'ing without a -f will only move trash further into the trash?

**Comment by keithj** [comment]
Yes, that's correct - we have trash off in this zone. The reason I'm using `irm` rather than `irmtrash` is because we have trash disabled, so I was searching for an alternative way of removing them.

Operating on a single file at a time always succeeded in removing files (without `-f`). The files that were unexpectedly left in place after the parallel `irm` runs had the same names as before the operation.

**Comment by tempoz** [comment]
I tried this with 30000 files and 20 threads and could not reproduce it. I irm'd the files from a collection in trash with the trash off. Do you have any additional information about the setup that might be relevant?

**Comment by tempoz** [comment]
I tried it again on oracle (instead of the postgres installation I naively  attempted at first) and I get errors about TNS listeners not being available for the operation requested, but no rmUtil errors. I still have the same problem you do, though, where an enormous number of files get skipped when this happens. I can make these go away by scaling back the number of threads passed to `parallel`. This looks like a limitation imposed by oracle on the number of connections it can accept. I do not know enough about oracle to know if that is a hard limit or something that can be configured. If you think I should change something about this setup to be able to reproduce your problem exactly, let me know; if, however, the workaround of just turning down the number of threads passed to `parallel` is acceptable to you and you do not wish to spend further time helping me reproduce this, I'll probably close this issue since the only thing I've been able to get to show up appears to be an oracle limitation.

**Comment by Pete-C** [comment]
Hello Zoe, Oracle can support connection pooling, depending upon which connector mechanism is used. This might be worth looking at as the team considers it's approach to supporting Oracle in the future.RegardsPete 

    On Tuesday, 20 June 2017, 22:21, Zoey Greer <notifications@github.com> wrote:
 

 I tried it again on oracle (instead of the postgres installation I naively attempted at first) and I get errors about TNS listeners not being available for the operation requested, but no rmUtil errors. I still have the same problem you do, though, where an enormous number of files get skipped when this happens. I can make these go away by scaling back the number of threads passed to parallel. This looks like a limitation imposed by oracle on the number of connections it can accept. I do not know enough about oracle to know if that is a hard limit or something that can be configured. If you think I should change something about this setup to be able to reproduce your problem exactly, let me know; if, however, the workaround of just turning down the number of threads passed to parallel is acceptable to you and you do not wish to spend further time helping me reproduce this, I'll probably close this issue since the only thing I've been able to get to show up appears to be an oracle limitation.—

**Comment by keithj** [comment]
I suspect that we do not have an issue with the number of TNS listeners here. Our workload is typically between 100 and 200 clients, 24x7, without observing significant numbers of failures. My 20 clients were well within tolerance of the numbers we normally operate successfully.

Having checked with our DBAs, we have 2 nodes, each of which can support up to 3000 sessions and each is currently operating at < 10% of that capacity. I think the TNS listeners limit you are hitting is specific to your own setup

**Comment by tempoz** [comment]
Thanks, guys; I'll see what I can do about making our oracle configuration we have in testing accept more load, then, and try this again.

**Comment by Pete-C** [comment]
Hello guys, there is a latency establishing each new connection. Connection pooling the should remove much of this latency, particularly under cluster based loads as this is also a potentially client tune-able. So yes the tmax qps looks fine, but there are potential benefits to be gained.
RegardsPete 

    On Wednesday, 21 June 2017, 15:48, Zoey Greer <notifications@github.com> wrote:
 

 Thanks, guys; I'll see what I can do about making our oracle configuration we have in testing accept more load, then, and try this again.—

**Comment by tempoz** [comment]
I have increased the number of processes our oracle installation could handle and still could not reproduce this error locally. Any further exploration of it would probably require you to repeat it on your system with increased logging and send us those logs.

**Comment by tempoz** [comment]
Is this also being done over federation (I assume yes)?

**Comment by trel** [comment]
@keithj @Pete-C Please send any captured logging for this case.  We're looking to finalize the work on 4-1-stable soon and we'd like to have this included if possible.

**Comment by trel** [comment]
The Oracle database connection is held by ODBC in 4.2+.  

Since we have not reproduced this in 4.1.x recently... I'm considering marking this as `wontfix`.
